sbin/dmctl/dmctl.8
.\"     $NetBSD: dmctl.8,v 1.5 2017/06/04 08:34:34 abhinav Exp $
.\"
.\" Copyright (c) 2004,2009 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Adam Hamsik.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.Dd September 11, 2016
.Dt DMCTL 8
.Os
.Sh NAME
.Nm dmctl
.Nd manipulate device-mapper driver command
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
.Nm
works with the device-mapper kernel driver.
It can send and receive information and commands from the dm driver.
.Bl -column -offset indent "suspend" "Switch active and passive tables for"
.It Li version Ta Print driver and lib version.
.It Li targets Ta List available kernel targets.
.It Li create Ta Create device with [dm device name].
.It Li ls Ta List existing dm devices.
.It Li info Ta Get info about device with [dm device name].
.It Li rename Ta Rename device with [dm device name] to [dm device new name].
.It Li remove Ta Remove device with [dm device name].
.It Li resume Ta Resume IO on dm device [dm device name].
.It Li suspend Ta Suspend IO on dm device [dm device name].
.It Li deps Ta Print physical dependencies for dm device [dm device name].
.It Li reload Ta Switch active and passive tables for [dm device name].
.It Li status Ta Print status for device with [dm device name].
.It Li table Ta Print active table for device with [dm device name].
.El
.Sh SEE ALSO
.Xr dm 3 ,
.Xr dm 4
.Sh HISTORY
The
.Nm
utility appeared in
.Nx 6.0 .
.Sh AUTHORS
.Nm
was written and contributed to
.Nx
by
.An Adam Hamsik .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2286.2.1 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

sbin/dmctl/dmctl.c
/*
 * Copyright (c) 2010 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Adam Hamsik.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/param.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <prop/proplib.h>

#include <dm.h>

#ifdef RUMP_ACTION
#include <rump/rump.h>
#include <rump/rumpclient.h>
#include <rump/rump_syscalls.h>
#endif

/* dmctl command is used to communicate with device-mapper driver in NetBSD
 * it uses libdm library to create and send required data to kernel.
 *
 * Main purpose of dmctl is to add posibility to use device-mapper driver
 * from outside of LVM scope.
 */

#define DMCTL_CMD_REQ_NODEVNAME 0	/* Command do not require device name */
#define DMCTL_CMD_REQ_DEVNAME	1	/* Command require device name to work */
#define DMCTL_CMD_REQ_NEWNAME	2	/* Command require device name and
					   newname to work */
struct command {
	const char *cmd_name;
	const char *cmd_help;
	const char *ioctl_cmd_name;
	int min_args;
	int (*cmd_func)(int, char *[], libdm_task_t);
};

static const   char *dvname;            /* device name */
static const   char *cmdname;           /* command user issued */

static char * parse_stdin(char *);

static int dmctl_get_version(int, char *[], libdm_task_t);
static int dmctl_get_targets(int, char *[], libdm_task_t);
static int dmctl_get_device_info(int, char *[], libdm_task_t);
static int dmctl_create_dev(int, char *[], libdm_task_t);
static int dmctl_dev_rename(int, char *[], libdm_task_t);
static int dmctl_dev_remove(int, char *[], libdm_task_t);
static int dmctl_dev_resume(int, char *[], libdm_task_t);
static int dmctl_dev_suspend(int, char *[], libdm_task_t);
static int dmctl_dev_deps(int, char *[], libdm_task_t);
static int dmctl_list_devices(int, char *[], libdm_task_t);
static int dmctl_table_reload(int, char *[], libdm_task_t);
static int dmctl_table_status(int, char *[], libdm_task_t);
__dead static void usage(void);

static struct command commands[] = {
	{ "version",
	  "Print driver and lib version.",
	  NULL, DMCTL_CMD_REQ_NODEVNAME,
	  dmctl_get_version },
	{ "targets",
	  "List available kernel targets.",
	  NULL, DMCTL_CMD_REQ_NODEVNAME,
	  dmctl_get_targets },
	{ "create",
	  "Create device with [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_create_dev },
	{ "ls",
	  "List existing dm devices.",
	  "names", DMCTL_CMD_REQ_NODEVNAME,
	  dmctl_list_devices },
	{ "info",
	  "Get info about device with [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_get_device_info },
	{ "rename",
	  "Rename device with [dm device name] to  [dm device new name].",
	  NULL, DMCTL_CMD_REQ_NEWNAME,
	  dmctl_dev_rename },
	{ "remove",
	  "Remove device with [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_dev_remove },
	{ "resume",
	  "Resume IO on dm device [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_dev_resume },
	{ "suspend",
	  "Suspend IO on dm device [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_dev_suspend },
	{ "deps",
	  "Print physical dependiences for dm device [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_dev_deps },
	{ "reload",
	  "Switch active and passive tables for device with [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_table_reload },
	{ "status",
	  "Print status for device with [dm device name].",
	  "table", DMCTL_CMD_REQ_DEVNAME,
	  dmctl_table_status },
	{ "table",
	  "Print active table for device with [dm device name].",
	  NULL, DMCTL_CMD_REQ_DEVNAME,
	  dmctl_table_status },
	{ NULL,
	  NULL,
	  NULL, 0,
	  NULL },
};

int
main(int argc, char *argv[])
{
	int i;
	int oargc;
	libdm_task_t task;

	oargc = 0;

#ifdef RUMP_ACTION
	if (rumpclient_init() == -1)
	  err(EXIT_FAILURE, "rump client init failed");
#endif

	/* Must have at least: device command */
	if (argc < 2)
		usage();

	/* Skip program name, get and skip device name and command. */
	cmdname = argv[1];
	if (argc > 2) {
		oargc = 1;
		dvname = argv[2];
	}

	if (argc > 3) {
		argv += 3;
		argc -= 3;
		oargc = 2;
	} else {
		argv = 0;
		argc = 0;
	}

	for (i = 0; commands[i].cmd_name != NULL; i++)
		if (strcmp(cmdname, commands[i].cmd_name) == 0)
			break;

	if (commands[i].cmd_name == NULL)
		errx(EXIT_FAILURE, "unknown command: %s", cmdname);

	if (commands[i].ioctl_cmd_name != NULL)
		cmdname = commands[i].ioctl_cmd_name;

	if (oargc != commands[i].min_args) {
		(void)fprintf(stderr, "Insufficient number of arguments for "
		    "command: %s specified\n", commands[i].cmd_name);
		usage();
	}

	/*
	 * Create libdm task, and pass it to command handler later.
	 * Don't release it here because it will be replaced by different
	 * dictionary received from kernel after libdm_task_run.
	 */
	task = libdm_task_create(cmdname);

	(*commands[i].cmd_func)(argc, argv, task);

	return 0;
}

/*
 * Print library and kernel driver versions if command can be used only when
 * major, minor number of library version is <= kernel.
 */
static int
dmctl_get_version(int argc __unused, char *argv[] __unused, libdm_task_t task)
{
	uint32_t ver[3];

	(void)libdm_task_get_cmd_version(task, ver, sizeof(ver));

	printf("Library protocol version %d:%d:%d\n", ver[0], ver[1], ver[2]);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_get_version: libdm_task_run failed.");

	(void)libdm_task_get_cmd_version(task, ver, 3);
	printf("Kernel protocol version %d:%d:%d\n",ver[0], ver[1], ver[2]);

	libdm_task_destroy(task);
	return 0;
}

/*
 * Get list of available targets from kernel and print them.
 */
static int
dmctl_get_targets(int argc __unused, char *argv[] __unused, libdm_task_t task)
{
	libdm_cmd_t cmd;
	libdm_iter_t iter;
	libdm_target_t target;
	uint32_t ver[3];

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_get_targets: libdm_task_run failed.");

	if ((cmd = libdm_task_get_cmd(task)) == NULL)
		return ENOENT;

	iter = libdm_cmd_iter_create(cmd);

	while((target = libdm_cmd_get_target(iter)) != NULL){
		printf("Target name: %s\n", libdm_target_get_name(target));

		libdm_target_get_version(target, ver, sizeof(ver));
		printf("Target version %d.%d.%d\n\n", ver[0], ver[1], ver[2]);

		libdm_target_destroy(target);
	}

	libdm_iter_destroy(iter);
	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);

	return 0;
}

/*
 * Create device with name used as second parameter.
 * TODO: Support for UUIDs here.
 */
static int
dmctl_create_dev(int argc __unused, char *argv[] __unused, libdm_task_t task)
{

	libdm_task_set_name(dvname, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_create_dev: libdm_task_run failed.");

	libdm_task_destroy(task);
	return 0;
}

/*
 * Get basic device info from device-mapper driver.
 */
static int
dmctl_get_device_info(int argc __unused, char *argv[] __unused, libdm_task_t task)
{

	libdm_task_set_name(dvname, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "%s: libdm_task_run failed", __func__);

	printf("Printing Device info for:\n");
	printf("Device name: \t\t%s\n", libdm_task_get_name(task));
	printf("Device uuid: \t\t%s\n", libdm_task_get_uuid(task));
	printf("Device minor: \t\t%d\n", libdm_task_get_minor(task));
	printf("Device target number: \t%d\n", libdm_task_get_target_num(task));
	printf("Device flags: \t\t%d\n", libdm_task_get_flags(task));

	libdm_task_destroy(task);
	return 0;
}

/*
 * List all device in device-mapper driver.
 */
static int
dmctl_list_devices(int argc __unused, char *argv[] __unused, libdm_task_t task)
{
	libdm_cmd_t cmd;
	libdm_iter_t iter;
	libdm_dev_t dev;

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_list_devices: libdm_task_run failed.");

	if ((cmd = libdm_task_get_cmd(task)) == NULL)
		return ENOENT;

	iter = libdm_cmd_iter_create(cmd);

	while((dev = libdm_cmd_get_dev(iter)) != NULL){
		printf("Device name: %s, device minor: %d \n",
		    libdm_dev_get_name(dev), libdm_dev_get_minor(dev));
		libdm_dev_destroy(dev);
	}

	libdm_iter_destroy(iter);
	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);

	return 0;
}

/*
 * Rename device to new name
 */
static int
dmctl_dev_rename(int argc __unused, char *argv[], libdm_task_t task)
{
	libdm_cmd_t cmd;

	libdm_task_set_name(dvname, task);

	cmd = libdm_cmd_create();
	libdm_dev_set_newname(argv[0], cmd);
	libdm_task_set_cmd(cmd, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_dev_rename: libdm_task_run failed.");

	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);

	return 0;
}

/*
 * Remove device from dm device list.
 */
static int
dmctl_dev_remove(int argc __unused, char *argv[] __unused, libdm_task_t task)
{

	if (dvname == NULL)
		return (ENOENT);

	libdm_task_set_name(dvname, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_dev_remove: libdm_task_run failed.");

	libdm_task_destroy(task);
	return 0;
}

/*
 * Resume device which was suspended or created right now.
 * Replace table in "active slot" witg table in "inactive slot".
 */
static int
dmctl_dev_resume(int argc __unused, char *argv[] __unused, libdm_task_t task)
{

	libdm_task_set_name(dvname, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_dev_resume: libdm_task_run failed.");

	libdm_task_destroy(task);
	return 0;
}

/*
 * Resume device which was suspended or created right now.
 * Replace table in "active slot" with table in "inactive slot".
 */
static int
dmctl_dev_suspend(int argc __unused, char *argv[] __unused, libdm_task_t task)
{

	libdm_task_set_name(dvname, task);
	libdm_task_set_suspend_flag(task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_dev_suspend: libdm_task_run failed.");

	libdm_task_destroy(task);
	return 0;
}

/*
 * Get device dependiences from device-mapper. Device dependency is physical
 * device on which dm device depends.
 */
static int
dmctl_dev_deps(int argc __unused, char *argv[] __unused, libdm_task_t task)
{
	libdm_cmd_t cmd;
	libdm_iter_t iter;
	dev_t dev_deps;

	libdm_task_set_name(dvname, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "dmctl_dev_deps: libdm_task_run failed.");

	if ((cmd = libdm_task_get_cmd(task)) == NULL)
		return ENOENT;

	iter = libdm_cmd_iter_create(cmd);

	printf("Device %s dependiences \n", dvname);

	while((dev_deps = libdm_cmd_get_deps(iter)) != 0)
		printf("major: %d minor: %d\n", major(dev_deps), minor(dev_deps));

	libdm_iter_destroy(iter);
	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);

	return 0;
}

/*
 * Reload device table to get new one to use.
 */
static int
dmctl_table_reload(int argc, char *argv[], libdm_task_t task)
{
	libdm_cmd_t cmd;
	libdm_table_t table;

	char *params;
	char *file_path;
	char target[128];
	int len;
	uint64_t start, length;

	file_path = NULL;
	params = NULL;

	cmd = libdm_cmd_create();

	libdm_task_set_name(dvname, task);

	if (argc != 0)
		file_path = argv[0];

	while ((params = parse_stdin(file_path)) != NULL) {
		table = libdm_table_create();

		sscanf(params, "%"PRIu64" %"PRIu64" %s %n", &start, &length, target, &len);

		libdm_table_set_start(start, table);
		libdm_table_set_length(length, table);
		libdm_table_set_target(target, table);
		libdm_table_set_params(params + len, table);
		libdm_cmd_set_table(table, cmd);

		libdm_table_destroy(table);

		free(params);
	}

	libdm_task_set_cmd(cmd, task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "libdm_task_run: from dmctl_table_reload failed.");

	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);
	free(params);

	return 0;
}

/*
 * Get table status from device.
 */
static int
dmctl_table_status(int argc __unused, char *argv[] __unused, libdm_task_t task)
{
	libdm_cmd_t cmd;
	libdm_table_t table;
	libdm_iter_t iter;

	libdm_task_set_name(dvname, task);
	libdm_task_set_status_flag(task);

	if (libdm_task_run(task) != 0)
		err(EXIT_FAILURE, "libdm_task_run");

	if ((cmd = libdm_task_get_cmd(task)) == NULL)
		return ENOENT;

	iter = libdm_cmd_iter_create(cmd);

	printf("Getting device table for device %s\n", dvname);

	while ((table = libdm_cmd_get_table(iter)) != NULL) {
		printf("%10"PRIu64" %10"PRIu64" %s\n",
			libdm_table_get_start(table),
			libdm_table_get_length(table),
			libdm_table_get_target(table));
		libdm_table_destroy(table);
	}

	libdm_iter_destroy(iter);
	libdm_cmd_destroy(cmd);
	libdm_task_destroy(task);

	return 0;
}

static void
usage(void)
{
	int i;

	(void)fprintf(stderr, "usage: %s command [dm device name]\n"
	    "Available commands are:\n ", getprogname());
	for (i = 0; commands[i].cmd_name != NULL; i++)
		(void)fprintf(stderr, "\t%s\t%s\n", commands[i].cmd_name, commands[i].cmd_help);
	exit(EXIT_FAILURE);
}

static char *
parse_stdin(char *file_path)
{
	char *buf, *lbuf;
	size_t len;
	FILE *fp;

	lbuf = NULL;

	if (file_path) {
		if ((fp = fopen(file_path, "r")) == NULL)
			err(ENOENT, "Cannot open table file");
	} else
		fp = stdin;

	if ((buf = fgetln(fp, &len)) == NULL)
		return NULL;

	if (buf[len - 1] != '\n')
		len++;

	if ((lbuf = malloc(len)) == NULL)
		err(EXIT_FAILURE, "malloc");

	memcpy(lbuf, buf, len);
	lbuf[len - 1] = '\0';

	return lbuf;
}
